### PR TITLE
disttask: fix wrongly add subtask when state transform fail

### LINKED ISSUE #46261: Disttask: task state can't transform from canceling to reverted.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
func TestFrameworkCancelThenSubmitSubTask(t *testing.T) {
	defer dispatcher.ClearTaskFlowHandle()
	defer scheduler.ClearSchedulers()
	var m sync.Map
	RegisterTaskMeta(&m)
	distContext := testkit.NewDistExecutionContext(t, 3)
	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/disttask/framework/dispatcher/cancelBeforeUpdate", "1*return(true)"))
	DispatchTaskAndCheckFail("😊", t, &m)
	require.NoError(t, failpoint.Disable("github.com/pingcap/tidb/disttask/framework/dispatcher/cancelBeforeUpdate"))
	distContext.Close()
}
```

### 2. What did you expect to see? (Required)
Task failed.
### 3. What did you see instead (Required)
Task hang.
### 4. What is your TiDB version? (Required)
master.

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: close #46261 ref #46258 

Problem Summary:
see #46261
1. When state transform fails, framework still adds subtasks.
2. can't transform state from cancelling to reverted.(If there are no subtasks for task submitted, this state transform is legal.)
### What is changed and how it works?
1. When state transform fails, check the affected row cnt to skip adding subtask. This error should not be retryable.
2. When there are no subtasks submitted and the task is canceled, the state transform will be canceling--->reverting--->reverted. Remove the old fast path: canceling--->reverted.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: disttask/framework/dispatcher/dispatcher.go [hunk 1]
```diff
@@ -219,9 +219,18 @@ func (d *dispatcher) updateTask(taskState string, newSubTasks []*proto.Subtask,
 		return errors.Errorf("invalid task state transform, from %s to %s", prevState, taskState)
 	}
 
+	failpoint.Inject("cancelBeforeUpdate", func(val failpoint.Value) {
+		if val.(bool) {
```
Comment (D3Hunter) [inline disttask/framework/dispatcher/dispatcher.go:223]: seems no need to check this var

File: disttask/framework/dispatcher/dispatcher.go [hunk 2]
```diff
@@ -454,6 +463,7 @@ func VerifyTaskStateTransform(oldState, newState string) bool {
 		proto.TaskStateRevertPending: {},
 		proto.TaskStateReverted:      {},
 	}
+	logutil.BgLogger().Info("task state transform", zap.String("oldState", oldState), zap.String("newState", newState))
```
Comment (D3Hunter) [inline disttask/framework/dispatcher/dispatcher.go:466]:
```suggestion
	logutil.BgLogger().Info("task state transform", zap.String("from", oldState), zap.String("to", newState))
```

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Pull request merged**